src/Xrefcheck/Verify.hs

```diff
@@ -231,22 +274,39 @@ data NeedsCaching key
 -- then the @action@ will not be executed, and the value is added to the accumulator list.
 -- After the whole list has been traversed, the accumulator is traversed once again to ensure
 -- every asynchronous action is completed.
+-- If interrupted by AsyncException, returns this exception and list of already calcualted results
+-- (their subset can be arbitrary). Computations that were not ended till this moment are cancelled.
 forConcurrentlyCaching
   :: Ord cacheKey
-  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
-forConcurrentlyCaching list needsCaching action = go [] M.empty list
+  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO (Either (AsyncException, [b]) [b])
+forConcurrentlyCaching list needsCaching action = do
+  res <- go [] M.empty list
+  case res of
+    Right r -> return $ Right $ map (fromMaybe (error "impossible")) r
+      -- impossible, since no actions were cancelled
+
+    Left (exception, partialResult) -> return $  Left (exception, catMaybes partialResult)
   where
-    go acc cached (x : xs) = case needsCaching x of
-      NoCaching -> do
-        withAsync (action x) $ \b ->
-          go (b : acc) cached xs
-      CacheUnderKey cacheKey -> do
-        case M.lookup cacheKey cached of
-          Nothing -> do
-            withAsync (action x) $ \b ->
-              go (b : acc) (M.insert cacheKey b cached) xs
-          Just b -> go (b : acc) cached xs
-    go acc _ [] = for acc wait <&> reverse
+    interruptibleAction a = handle (\AsyncCancelled -> pure Nothing) (Just <$> action a)
+
+    go acc cached items = handle
+      (\exception -> Left . (exception,) <$> for acc (\f -> cancel f >> wait f))
```

Comment:
I think this would be easier to read if we (1) avoided point-free notation, (2) used `do` blocks, and (3) renamed `f` to `asyncAction`.

```suggestion
      (\exception -> do
          partialResults <- for acc \asyncAction -> do
            cancel asyncAction
            wait asyncAction
          pure $ Left (exception, partialResults)
      )
```